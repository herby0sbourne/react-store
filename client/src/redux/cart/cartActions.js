import cartActionTypes from './cartTypes';

export const toggleCartDropdown = () => {
  return {
    type: cartActionTypes.TOGGLE_CART_DROPDOWN,
  };
};

export const addToCart = (item) => {
  return {
    type: cartActionTypes.ADD_ITEM,
    item,
  };
};

export const reduceItem = (item) => {
  return {
    type: cartActionTypes.REDUCE_ITEM,
    item,
  };
};

export const removeFromCart = (id) => {
  return {
    type: cartActionTypes.REMOVE_ITEM,
    id,
  };
};

export const emptyCart = () => {
  return {
    type: cartActionTypes.EMPTY_CART,
  };
};

export const updateCartInFirebase = () => {
  return {
    type: cartActionTypes.UPDATE_CART_IN_FIREBASE,
  };
};

export const setCartFromFirebase = (cartItems) => {
  return {
    type: cartActionTypes.SET_CART_FROM_FIREBASE,
    cartItems,
  };
};
