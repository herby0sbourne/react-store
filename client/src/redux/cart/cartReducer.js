import cartActionTypes from './cartTypes';
import { addItemToCart, removeItemFromCart } from './cartUtils';

const DROPDOWN_DEFAULT = {
  hidden: false,
  cartItems: [],
};

const cartReducer = (state = DROPDOWN_DEFAULT, action) => {
  switch (action.type) {
    case cartActionTypes.TOGGLE_CART_DROPDOWN:
      return {
        ...state,
        hidden: !state.hidden,
      };

    case cartActionTypes.ADD_ITEM:
      return {
        ...state,
        cartItems: addItemToCart(state.cartItems, action.item),
        // cartItems: [...state.cartItems, action.item],
      };

    case cartActionTypes.REDUCE_ITEM:
      return {
        ...state,
        cartItems: removeItemFromCart(state.cartItems, action.item),
      };

    case cartActionTypes.REMOVE_ITEM:
      return {
        ...state,
        cartItems: state.cartItems.filter((cartItem) => cartItem.id !== action.id),
      };

    case cartActionTypes.EMPTY_CART:
      return {
        ...state,
        cartItems: [],
      };

    case cartActionTypes.SET_CART_FROM_FIREBASE:
      return {
        ...state,
        cartItems: action.cartItems,
      };

    default:
      return state;
  }
};

export default cartReducer;
