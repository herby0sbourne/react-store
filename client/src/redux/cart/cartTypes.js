const cartActionTypes = {
  TOGGLE_CART_DROPDOWN: 'TOGGLE_CART_DROPDOWN',
  ADD_ITEM: 'ADD_ITEM',
  REDUCE_ITEM: 'REDUCE_ITEM',
  REMOVE_ITEM: 'REMOVE_ITEM',
  EMPTY_CART: 'EMPTY_CART',
  SET_CART_FROM_FIREBASE: 'SET_CART_FROM_FIREBASE',
  UPDATE_CART_IN_FIREBASE: 'UPDATE_CART_IN_FIREBASE',
};

export default cartActionTypes;
