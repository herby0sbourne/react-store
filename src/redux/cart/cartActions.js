import cartActionTypes from './cartTypes';

export const toggleCartDropdown = () => {
  return {
    type: cartActionTypes.TOGGLE_CART_DROPDOWN,
  };
};

export const addToCart = (item) => {
  return {
    type: cartActionTypes.ADD_ITEM,
    item,
  };
};

export const reduceItem = (item) => {
  return {
    type: cartActionTypes.REDUCE_ITEM,
    item,
  };
};

export const removeFromCart = (id) => {
  return {
    type: cartActionTypes.REMOVE_ITEM,
    id,
  };
};

export const emptyCart = () => {
  return {
    type: cartActionTypes.EMPTY_CART,
  };
};
