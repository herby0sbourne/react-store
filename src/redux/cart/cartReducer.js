import cartActionTypes from './cartTypes';
import { addItemToCart } from './cartUtils';

const DROPDOWN_DEFAULT = {
  hidden: false,
  cartItems: [],
};

const cartReducer = (state = DROPDOWN_DEFAULT, action) => {
  switch (action.type) {
    case cartActionTypes.TOGGLE_CART_DROPDOWN:
      return {
        ...state,
        hidden: !state.hidden,
      };

    case cartActionTypes.ADD_ITEM:
      return {
        ...state,
        cartItems: addItemToCart(state.cartItems, action.item),
        // cartItems: [...state.cartItems, action.item],
      };

    default:
      return state;
  }
};

export default cartReducer;
