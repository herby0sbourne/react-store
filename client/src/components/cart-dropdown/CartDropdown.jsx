import React from 'react';
import { connect } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import CartItem from '../cart-item/CartItem';
import CustomButton from '../custom-button/CustomButton';
import { selectCartItems } from '../../redux/cart/cartSelector';
import { toggleCartDropdown } from '../../redux/cart/cartActions';

import './cart-dropdown.scss';

const CartDropdown = ({ cartItems, dispatch }) => {
  const navigate = useNavigate();
  return (
    <div className="cart-dropdown">
      <div className="cart-items">
        {cartItems.length ? (
          cartItems.map((cartItem) => {
            return <CartItem key={cartItem.id} item={cartItem} />;
          })
        ) : (
          <span className="empty-message">Your cart is empty</span>
        )}
      </div>
      <CustomButton
        onClick={() => {
          navigate('/checkout');
          dispatch(toggleCartDropdown());
        }}
      >
        GO TO CHECKOUT
      </CustomButton>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    cartItems: selectCartItems(state),
  };
};

export default connect(mapStateToProps)(CartDropdown);
