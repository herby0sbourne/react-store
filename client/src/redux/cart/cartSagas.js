import { takeLatest, put, call, all } from 'redux-saga/effects';
import userActionTypes from '../user/userTypes';
import { emptyCart } from './cartActions';

export function* emptyCartOnSignOut() {
  yield put(emptyCart());
}

export function* onSignOutSuccess() {
  yield takeLatest(userActionTypes.SIGN_OUT_SUCCESS, emptyCartOnSignOut);
}

export function* cartSaga() {
  yield all([call(onSignOutSuccess)]);
}
