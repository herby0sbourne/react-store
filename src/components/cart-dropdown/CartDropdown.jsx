import React from 'react';
import { connect } from 'react-redux';
import CartItem from '../cart-item/CartItem';
import CustomButton from '../custom-button/CustomButton';

import './cart-dropdown.scss';

const CartDropdown = ({ cartItems }) => {
  return (
    <div className="cart-dropdown">
      <div className="cart-items">
        {cartItems.map((cartItem) => {
          // console.log(cartItem);
          return <CartItem key={cartItem.id} item={cartItem} />;
        })}
      </div>
      <CustomButton>GO TO CHECKOUT</CustomButton>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    cartItems: state.cart.cartItems,
  };
};

export default connect(mapStateToProps)(CartDropdown);
