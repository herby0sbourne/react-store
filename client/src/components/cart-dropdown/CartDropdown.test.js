import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CartDropdown } from './CartDropdown';
import { toggleCartDropdown } from '../../redux/cart/cartActions';

const mockedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigate,
}));

let mockDispatch;
mockDispatch = jest.fn();

test('should render cart dropdown with empty message', () => {
  expect.assertions(1);
  const wrapper = shallow(<CartDropdown cartItems={[]} dispatch={mockDispatch} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('should render cart dropdown with with cart item', () => {
  expect.assertions(1);
  const wrapper = shallow(<CartDropdown cartItems={cartItems} dispatch={mockDispatch} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('should call navigate when button is clicked', () => {
  expect.assertions(2);
  const wrapper = shallow(<CartDropdown cartItems={[]} dispatch={mockDispatch} />);
  wrapper.find('CustomButton').simulate('click');
  expect(mockedNavigate).toHaveBeenCalledWith('/checkout');
  expect(mockDispatch).toHaveBeenCalledWith(toggleCartDropdown());
});

const cartItems = [
  {
    id: 1,
    imageUrl: 'https://www.w3schools.com/w3css/tryit.asp?filename=tryw3css_default_intro',
    price: '1200',
    name: 'New Car',
    quantity: '3',
  },
  {
    id: 2,
    imageUrl:
      'https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/gh-best-washing-machines-1589999096.png?crop=0.654xw:1.00xh;0.175xw,0&resize=640:*',
    price: '100',
    name: 'washing Machine',
    quantity: '1',
  },
];
