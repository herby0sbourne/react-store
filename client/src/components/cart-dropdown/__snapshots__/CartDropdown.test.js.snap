// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render cart dropdown with empty message 1`] = `
<div
  className="cart-dropdown"
>
  <div
    className="cart-items"
  >
    <span
      className="empty-message"
    >
      Your cart is empty
    </span>
  </div>
  <CustomButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CustomButton>
</div>
`;

exports[`should render cart dropdown with with cart item 1`] = `
<div
  className="cart-dropdown"
>
  <div
    className="cart-items"
  >
    <Memo(CartItem)
      item={
        Object {
          "id": 1,
          "imageUrl": "https://www.w3schools.com/w3css/tryit.asp?filename=tryw3css_default_intro",
          "name": "New Car",
          "price": "1200",
          "quantity": "3",
        }
      }
      key="1"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 2,
          "imageUrl": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/gh-best-washing-machines-1589999096.png?crop=0.654xw:1.00xh;0.175xw,0&resize=640:*",
          "name": "washing Machine",
          "price": "100",
          "quantity": "1",
        }
      }
      key="2"
    />
  </div>
  <CustomButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CustomButton>
</div>
`;
